### PR TITLE
split: skip zero len writes

### PR DESCRIPTION
**Pull request by ignoramous**
I'm not sure why (as zero len writes are apparently no-op in Golang) but the Quad9 resolver (DNSCrypt over TCP) doesn't work without this check.

### CODE REVIEW

File: Android/app/src/go/intra/split/retrier.go [hunk 1]
```diff
@@ -318,14 +318,21 @@ func (r *retrier) Write(b []byte) (int, error) {
...
 			r.mutex.Lock()
 			r.mutex.Unlock()
-			m, err := r.conn.Write(b[n:])
-			return n + m, err
+	
+			if len(leftover) > 0 {
+				m, err := r.conn.Write(leftover)
+				return n + m, err
+			}
+			return n, nil
```
Comment (jyyi1) [inline Android/app/src/go/intra/split/retrier.go:335]:
```suggestion
			if buf := b[n:]; len(buf) > 0 {
				m, e := r.conn.Write(buf)
				n += m
				err = e
			}
			return n, err
```

File: Android/app/src/go/intra/split/retrier.go [hunk 2]
```diff
@@ -318,14 +318,21 @@ func (r *retrier) Write(b []byte) (int, error) {
 			if err == nil {
 				return n, nil
 			}
+
+			leftover := b[n:]
```
Comment (jyyi1) [inline Android/app/src/go/intra/split/retrier.go:322]:
```suggestion
```

File: Android/app/src/go/intra/split/retrier.go [hunk 3]
```diff
@@ -318,14 +318,21 @@ func (r *retrier) Write(b []byte) (int, error) {
...
 			// A write error occurred on the provisional socket.  This should be handled
 			// by the retry procedure.  Block until we have a final socket (which will
 			// already have replayed b[:n]), and retry.
 			<-r.retryCompleteFlag
+	
 			r.mutex.Lock()
 			r.mutex.Unlock()
-			m, err := r.conn.Write(b[n:])
-			return n + m, err
+
```
Comment (jyyi1) [inline Android/app/src/go/intra/split/retrier.go:330]: Please also add a comment here to explain the reason we are skipping 0-len buffers.

**Review by jyyi1 — CHANGES_REQUESTED** [review]

File: Android/app/src/go/intra/split/retrier.go (hunk 1 above)
Comment (ignoramous) [reply Android/app/src/go/intra/split/retrier.go]: Nice. Thanks.

**Review by jyyi1 — APPROVED** [review]

**Comment by Kjs7945** [comment]
> I'm not sure why (as zero len writes are apparently no-op in Golang) but the Quad9 resolver (DNSCrypt over TCP) doesn't work without this check.

**Review by dekwan44 — APPROVED** [review]

**Pull request merged**